exercises/robot-name/examples/success-standard/src/Robot.hs

```diff
@@ -1,22 +1,43 @@
-module Robot (Robot, mkRobot, robotName, resetName) where
+module Robot (Robot, mkRobot, resetName, robotName, initialState) where
 
-import Control.Concurrent (MVar, readMVar, swapMVar, newMVar)
-import Control.Monad (void)
-import System.Random (randomRIO)
+import           Control.Concurrent.MVar (MVar, newMVar, putMVar, readMVar,
+                                          takeMVar)
+import           Control.Monad.State     (StateT, get, modify)
+import           Control.Monad.Trans     (lift)
+import           Data.Set                (Set)
+import qualified Data.Set                as S
+import           System.Random           (randomRIO)
 
 newtype Robot = Robot { robotNameVar :: MVar String }
+type RunState = Set String
 
-mkRobot :: IO Robot
-mkRobot = fmap Robot $ generateName >>= newMVar
+initialState :: RunState
+initialState = S.empty
+
+randomName :: IO String
+randomName = mapM randomRIO [letter, letter, digit, digit, digit]
+  where
+    letter = ('A', 'Z')
+    digit = ('0', '9')
+
+randomUniqueName :: StateT RunState IO String
+randomUniqueName = do
+  name <- lift randomName
+  nameSpace <- get
+  if name `S.member` nameSpace
+    then randomUniqueName
+    else return name
```

Comment:
I may be missing something here, and it's quite likely(!), but were does the new name get stored back into the set from initialState?